migration/migration2.go

```diff
@@ -0,0 +1,189 @@
+		if err := v2MigrationDealIDFilling(ds, propDealIDMap); err != nil {
+			return fmt.Errorf("migrating job logger: %s", err)
+		}
+		log.Infof("DealID filling migration finished")
+
+		return nil
+	},
+}
+
+func v2GenProposalCidToDealID(ds datastoreReaderWriter) (map[cid.Cid]uint64, error) {
+	ret := map[cid.Cid]uint64{}
+	qs := []query.Query{{Prefix: "/deals/storage-pending"}, {Prefix: "/deals/storage-final"}}
+
+	for _, q := range qs {
+		var lock sync.Mutex
+		var errors []string
+		lim := make(chan struct{}, 1000)
+
+		res, err := ds.Query(q)
+		if err != nil {
+			return nil, fmt.Errorf("querying storage info: %s", err)
+		}
+		defer func() {
+			_ = res.Close()
+		}()
+
+		for r := range res.Next() {
+			if r.Error != nil {
+				return nil, fmt.Errorf("query result: %s", r.Error)
+			}
+			lim <- struct{}{}
+
+			r := r
+			go func() {
+				defer func() { <-lim }()
+				var dr deals.StorageDealRecord
+				if err := json.Unmarshal(r.Value, &dr); err != nil {
+					lock.Lock()
+					errors = append(errors, fmt.Sprintf("unmarshaling query result: %s", err))
+					lock.Unlock()
+					return
+				}
+				if dr.DealInfo.DealID != 0 {
+					lock.Lock()
+					ret[dr.DealInfo.ProposalCid] = dr.DealInfo.DealID
+					lock.Unlock()
+				}
+			}()
+		}
+		for i := 0; i < cap(lim); i++ {
+			lim <- struct{}{}
+		}
+
+		if len(errors) > 0 {
+			for _, m := range errors {
+				log.Error(m)
+			}
+			return nil, fmt.Errorf("map build had %d errors", len(errors))
+		}
+	}
+
+	return ret, nil
+}
+
+type v1PartialStorageInfo struct {
+	Cold v1PartialCold
+}
+
+type v1PartialCold struct {
+	Filecoin v1PartialFilecoinInfo
+}
+
+type v1PartialFilecoinInfo struct {
+	Proposals []v1PartialFilStorage
+}
+
+type v1PartialFilStorage struct {
+	ProposalCid cid.Cid
+}
+
+func v2MigrationDealIDFilling(ds datastoreReaderWriter, propsDealID map[cid.Cid]uint64) error {
+	log.Infof("Total map size: %d", len(propsDealID))
+	q := query.Query{Prefix: "/ffs/scheduler/cistore_v2"} // /ffs/scheduler/cistore_v2/<iid>/<cid>
+	res, err := ds.Query(q)
+	if err != nil {
+		return fmt.Errorf("querying storage info store: %s", err)
+	}
+	defer func() { _ = res.Close() }()
+
+	var lock sync.Mutex
+	var errors []string
+	lim := make(chan struct{}, 1000)
+
+	var cantBeMapped, canBeMapped int
+	for r := range res.Next() {
+		if r.Error != nil {
+			return fmt.Errorf("iterating result: %s", r.Error)
+		}
+		lim <- struct{}{}
+
+		r := r
+		go func() {
+			defer func() { <-lim }()
+
+			var oldCi v1PartialStorageInfo
+			var newCi ffs.StorageInfo
+			if err := json.Unmarshal(r.Value, &oldCi); err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("unmarshaling storage info from datastore: %s", err))
+				lock.Unlock()
+				return
+			}
+			if err := json.Unmarshal(r.Value, &newCi); err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("unmarshaling storage info from datastore: %s", err))
+				lock.Unlock()
+				return
+			}
```

Comment:
A very interesting situation here: why unmarshaling two times?

The problem is the following. In this version of the code `ffs.StorageInfo` already has wiped out `ProposalCid`. So the unmarshaling using that type won't have the `ProposalCid` we need for taking a look at the `ProposalCid->DealID` map!

So I created `v1PartialStorageInfo` which is basically a partial struct of the old (current) JSON with the _ProposalCid_ field. Then we also unmashal the same JSON but with the new `ffs.StorageInfo` to fill the _DealID_ field.